src/main/kotlin/org/jitsi/videobridge/util/ArrayTracker.kt

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright @ 2019 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.videobridge.util
+
+import java.lang.Integer.max
+import org.jitsi.nlj.util.floorMod
+import java.lang.Integer.min
+
+/**
+ * Implements a fixed-sized tracker based on a pre-filled array. The main use-case is the frame projection tracker.
+ *
+ * @author Jonathan Lennox, based on [org.jitsi.nlj.util.ArrayCache] by Boris Grozev
+ */
+open class ArrayTracker<T>(
+    val size: Int
+) {
+    private val cache: Array<Container> = Array(size) { Container() }
+    protected val syncRoot = Any()
+    /**
+     * The index in [cache] where the item with the highest index is stored.
+     */
+    private var head = -1
+
+    protected val lastIndex: Int
+        get() = if (head == -1) -1 else cache[head].index
+
+    protected var firstIndex: Int = -1
```

Comment:
Could you add some docs to `firstIndex`, and `lastIndex`?

Reply:
Ok.